### ISSUE TITLE
Paper 11/13 — Related Work

### ISSUE THREAD

**Issue by murillo128** [murillo128/k3-out-of-core#120]
Parent: #109

**Editorial status:** `EVIDENCE_CHECK`

This issue coordinates Related Work only. It is not an execution contract.

## Purpose

Position the paper against the closest systems and routing prior art without turning Related Work into a bibliography dump. Prefer three conceptual groups and state the exact distinction from this work in each.

### 10.1 Expert offloading, caching, and local MoE serving

Primary references from `docs/PRIOR_ART.md`:

- MoE-Infinity — activation-aware expert caching/prefetch for personal machines;
- WASTE — full 2.78T K3 streamed from NVMe with custom 3-bit expert representation and measured I/O/cache limits;
- Colibrì — full text K3 with source-MXFP4 experts, direct I/O, CPU/Vulkan tiers, and chunked prefill;
- FreeToken — RAM-resident expert pool, global VRAM LRU, bandwidth-adaptive CPU/GPU miss execution;
- llama.cpp expert-cache/offload proposals and vLLM ExpertWeightProvider architecture where useful;
- tinyserve / other directly relevant cache implementations if space permits.

Core positioning:

> These systems primarily optimize **where and how exact router-selected experts are materialized or executed**. Our work additionally studies whether a small, explicitly bounded change in selected membership can reduce the demand presented to the out-of-core memory system.

Important distinctions:

```text
FreeToken: host-RAM expert pool -> VRAM cache / CPU execution
this work: NVMe-backed expert pool -> bounded host residency, with routing itself used as a bounded control
```

WASTE and Colibrì are stronger same-model/full-K3 context than FreeToken, but their representations/hardware differ enough that raw TPS must remain qualitative rather than directly normalized unless protocols truly align.

### 10.2 Cache-aware / residency-aware routing

Must include:

- **Mixture of Cache-Conditional Experts for Efficient Mobile Device Inference** — training-free cache-aware expert membership/reranking; direct conceptual prior art.
- **MoE-ERAS: Expert Residency Aware Selection** — historical residency-aware performance/accuracy trade-off.
- **ReMoE** — router fine-tuning to increase temporal reuse under memory constraints.

Do not claim this paper invents cache-aware MoE routing.

Safe differentiation:

- full K3 896-expert / top-16 setting;
- hard per-swap router-score regret bound;
- hard swap-count bound;
- deterministic training-free inference-time substitutions;
- original expert weighting semantics retained;
- physical NVMe-backed full-K3 evaluation;
- real downstream route evolution after substitutions;
- long-horizon direct-vs-free trajectory quality attribution.

### 10.3 Routing locality, specialization, and expert structure

Include as needed:

- Local Routing Consistency — locality/cacheability varies by model, motivating K3-specific measurement;
- PipeNetwork K3 REAP/domain-overlap work — K3-specific specialization/saliency prior, but not identical selected-route/cache observable;
- CommitteeAudit / Standing Committee — shared core + task-specific periphery prior on other MoEs.

Core positioning:

> We use these works to motivate route-structure analysis, but do not transfer their model-specific functional or numerical conclusions to K3.

## Suggested Related Work flow/language

For each group:

1. one sentence stating the shared problem;
2. 2-4 concise prior-work examples;
3. one explicit paragraph stating what this paper adds or measures differently.

Avoid one paragraph per paper unless a work is especially close.

## Prior-art language boundaries

- no `first` claim for cache-aware routing;
- no direct raw-TPS comparison across WASTE/Colibrì/FreeToken when model representation/hardware/protocol differ;
- do not describe external project results as if independently reproduced unless they were;
- distinguish `same model`, `same representation`, `same hardware`, and `same protocol` rather than collapsing all into "comparable";
- preserve #105's prior-art classifications when using its curated comparison.

## Source material

The durable project prior-art source is `docs/PRIOR_ART.md`, including the detailed FreeToken note and the K3-specific WASTE/Colibrì comparisons. Re-check current publication metadata before final submission, but do not expand the paper into a general MoE survey.

## Draft result — 2026-08-20

Related Work is drafted in `paper/paper.md` on `paper/manuscript-v0` at commit `098f16bc645889f48079b934cd935c4648d8def2`.

### Editorial decisions retained

- Keep exactly three conceptual groups: expert offloading/caching; cache-aware/residency-aware routing; routing locality/specialization.
- Treat MoE-Infinity, WASTE, Colibrì, FreeToken, llama.cpp/vLLM, and tinyserve as systems/architecture lineage rather than novelty. WASTE and Colibrì are same-model K3 context, but their representation, hardware, cache hierarchy, and protocols differ; their raw TPS is not used as a direct baseline. FreeToken is adjacent RAM-resident exact-routing context rather than a same-regime K3 comparator.
- State explicitly that cache-aware expert membership predates this work. Cache-Conditional Experts (TMLR 2025), MoE-ERAS (MLArchSys 2024), and ReMoE (ICML 2026) define the closest routing design space.
- Differentiate this paper through exact-route-relative hard per-swap K3 selection-score regret, hard `max_swaps`, deterministic training-free substitutions, preserved K3 weighting semantics, physical full-K3 NVMe-backed evaluation, real subsequent route evolution, and controlled long-horizon fixed-context versus free-trajectory quality attribution.
- Use Local Routing Consistency, PipeNetwork K3 REAP, and CommitteeAudit / Standing Committee to motivate K3-specific route-structure questions only. Do not transfer external overlap values or functional labels to K3. The §9 K3 core/periphery result remains a selected-route `POST_HOC_EXPLORATORY` analogue, not an exact reproduction of routing-weight-profile analyses.
- Publication metadata checked on 2026-08-20: Cache-Conditional Experts is published in TMLR (06/2025); MoE-ERAS in MLArchSys 2024; ReMoE is listed at ICML 2026; CommitteeAudit / Standing Committee is published at ACL 2026; FreeToken remains a recent arXiv preprint. Preserve the final-submission metadata recheck gate rather than treating this check as permanent.

### Reviewer-risk boundaries

- no first/priority claim for cache-aware routing;
- no direct heterogeneous raw-TPS comparison;
- no external result described as independently reproduced;
- no semantic-safety claim inferred from local router-score regret;
- no transfer of model-specific expert functions or numerical locality conclusions to K3.
